feature: added support for ani-skip to skip episode intros

AzureOrange404 opened the pull request:
# Pull Request

## Type of change

- [ ] Bug fix
- [x] Feature
- [ ] Documentation update

## Description

I saw the issue [#1223](https://github.com/pystardust/ani-cli/issues/1223) and sympathized with the idea of skipping intros.
So implemented the skipping function into ani-cli (it still needs ani-skip to be installed though.)

Skipping does not work for all anime I've tested, but does for some.
If ani-skip fails, ani-cli just prints a message stating, that it failed and proceeds as normal.

Also it only works with mpv (due to how ani-skip works) and this is noted in the help function as well as the man page.

## Checklist

- [x] any anime playing
- [x] bumped version
---
- [x] next, prev and replay work
- [x] `-c` history and continue work
- [x] `-d` downloads work
- [x] `-s` syncplay works
- [x] `-q` quality works
- [x] `-v` vlc works
- [x] `-e` select episode works
- [x] `-S` select index works
- [x] `-r` range selection works
- [x] `--dub` and regular (sub) mode both work
- [x] all providers return links (not necessarily on a single anime, use debug mode to confirm)
---
- [x] `-h` help info is up to date
- [x] Readme is up to date
- [x] Man page is up to date

## Additional Testcases

- The safe bet: One Piece
- Episode 0: Saenai Heroine no Sodatekata ♭
- Unicode: Saenai Heroine no Sodatekata ♭
- Non-whole episodes: Tensei shitara slime datta ken (ep. 24.5, ep. 24.9)

Derisis13 commented on ani-cli:
```diff
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-version_number="4.6.2"
+version_number="4.6.3"
```
Since this is a feature update, it could be a version bump to 4.7

AzureOrange404 replied:
done :)

Derisis13 commented on ani-cli:
```diff
@@ -370,6 +386,9 @@ done
 dep_ch "curl" "sed" "grep" || true
+if [ "$skip_intro" = 1 ]; then
+    dep_ch "ani-skip" || true
+fi
```
could this be consolidated to one line or does shellecheck complain then?

AzureOrange404 replied:
It did complain, but I realised I could  put the second statement into brackets – so it's done.

AzureOrange404 replied:
Can this be resolved?

Derisis13 reviewed (changes requested):
You did a nice job implementing the feature in a way that's consistent to the other parts of the script. There's just one thing missing: ani-skip install instructions in the readme. There should be a reference to their instructions with some description on what it does

port19x reviewed (commented):
Very nice implementation.
I have nothing to add to the review by @Derisis13

synacktraa commented:
> You did a nice job implementing the feature in a way that's consistent to the other parts of the script. There's just one thing missing: ani-skip install instructions in the readme. There should be a reference to their instructions with some description on what it does

Hi, I am opening a pull request to include the ani-skip directly into `ani-cli`. There's no need to include ani-skip as a dependency.

synacktraa commented:
I've also wrote a lua script for vlc player.

lasersPew commented:
Before you accept this pr, it would be better if you add ani-skip onto the features and the dependency requirements. Additionally, warn windows users as I get `ani-skip not installed` and there's no official app for it.

synacktraa commented:
> Before you accept this pr, it would be better if you add ani-skip onto the features and the dependency requirements. Additionally, warn windows users as I get `ani-skip not installed` and there's no official app for it.

I'll include the code directly into ani-cli script, you won't have to install ani-skip explicitly

Derisis13 commented:
> Hi, I am opening a pull request to include the ani-skip directly into `ani-cli`. There's no need to include ani-skip as a dependency.
> 
> It would be nice, If this PR is merged first. I'll only have to replace the function.

I'm against including lines of code instead of an available solution: just as we outsourced our ui to fzf and rofi, I see no point in re-implementing ani-skip. The reason is we have to maintain our code, but dependencies take care of themselves.

This PR looks fine as is.

Derisis13 commented:
> I've also wrote a lua script for vlc player. VLC is not as flexible as MPV, We'll need to find a suitable way to pass in options or hard-code them.

This is welcome once the PR is merged

AzureOrange404 commented:
> You did a nice job implementing the feature in a way that's consistent to the other parts of the script. There's just one thing missing: ani-skip install instructions in the readme. There should be a reference to their instructions with some description on what it does

I did update the README as well, I've put everything under the dependencies section (should be fine I guess).
Sorry about the thousend commits on the readme, it's definitely to late to work now ... haha.

> Before you accept this pr, it would be better if you add ani-skip onto the features and the dependency requirements. Additionally, warn windows users as I get `ani-skip not installed` and there's no official app for it.

 I added a warning to the ani-skip description. I should work under Android and MacOS right? (Not shure about iOS though.)

port19x commented:
I can test mac compatibility tomorrow, code and readme looks good.
I'm also against inclusion of ani-skip code in mainline, unless the expansion over the currently proposed solution would be negligible.
Either way I'm in favour of merging early and making concessions to windows users later.

Derisis13 commented:
> I'm also against inclusion of ani-skip code in mainline, unless the expansion over the currently proposed solution would be negligible

it's 27 lines of shell scripting and another 27 lines of lua. Since we can't include the lua scrip anyway I'd leave ani-skip in the hands of its current maintainer.

synacktraa commented:
> I can test mac compatibility tomorrow, code and readme looks good. I'm also against inclusion of ani-skip code in mainline, unless the expansion over the currently proposed solution would be negligible. Either way I'm in favour of merging early and making concessions to windows users later.

I've condensed the code into a single function, you guys can review it once.
```sh
get_skip_times() {
        name=$(printf "%s" "$title" | sed -E "s@(.+\b)( \().*@\1@")
        keyword=$(printf "%s" "$name" | sed -E 's| |%20|g')
        mal_output=$(curl -s "https://myanimelist.net/search/prefix.json?type=anime&keyword=$keyword")
        mal_id=$(printf "%s\n" "$mal_output" | grep -Eo '"id"[^}]+'| sed -E 's@"id":([0-9]{1,}),.+name":"([^"]+).+@\2|\1@g'| grep -Eo "^$name\|.*" | sed -E "s@.+\|(.+)@\1@")
        [ -z "$mal_id" ] && printf "\033[1;31mMyAnimeList ID not found!\033[0m\n" && return 1

        aniskip_output=$(curl -s "https://api.aniskip.com/v1/skip-times/$mal_id/$ep_no?types[]=op")
        found_status=$(printf "%s" "$aniskip_output" | sed -E 's_\{"found":(true|false),.*_\1_p' | head -n1)
        [ "$found_status" = "false" ] && printf "\033[1;31mSkip times not found!\033[0m\n" && return 1

        opts_flag=$(printf "%s" "$aniskip_output" | sed -E 's@.*"([a-z_]+)":([0-9.]+),"([a-z_]+)":([0-9.]+).*:"op".*@--script-opts=skip-\1=\2,skip-\3=\4@')
        skip_flag="$opts_flag --script=~/.config/mpv/scripts/skip.lua"
}
```
`$skip_flag` will be passed to mpv.

synacktraa commented:
> > I'm also against inclusion of ani-skip code in mainline, unless the expansion over the currently proposed solution would be negligible
> 
> it's 27 lines of shell scripting and another 27 lines of lua. Since we can't include the lua scrip anyway I'd leave ani-skip in the hands of its current maintainer.

Yes, lua scripts are another issue.

71zenith commented:
Pls add the copying of `ani-skip` shell script to `$PATH` in the readme install instructions

synacktraa commented:
> Pls add the copying of `ani-skip` shell script to `$PATH` in the readme install instructions

I have updated it, do tell me If there are any more issues with `ani-skip`.

synacktraa commented on ani-cli:
```diff
@@ -237,13 +250,14 @@ download() {
 }
 
 play_episode() {
+    [ "$skip_intro" = 1 ] && get_skip_times
```
Change it to  `[ "$skip_intro" = 1 ] && skip_flag="$(ani-skip "$title" "$ep_no")"`

AzureOrange404 replied:
changed to:
```sh
[ "$skip_intro" = 1 ] && skip_flag="$(ani-skip "$title" "$ep_no" | cut -d' ' -f1)" && skip_arg="$(ani-skip "$title" "$ep_no" | cut -d' ' -f2)"
```

synacktraa commented on ani-cli:
```diff
@@ -237,13 +250,14 @@ download() {
             printf "%s\n" "$episode"
             ;;
-        mpv*) nohup "$player_function" --force-media-title="${allanime_title}Episode ${ep_no}" "$episode" >/dev/null 2>&1 & ;;
+        mpv*) nohup "$player_function" "$skip_script" "$skip_argument" --force-media-title="${allanime_title}Episode ${ep_no}" "$episode" >/dev/null 2>&1 & ;;
```
Replace `"$skip_script" "$skip_argument"` with `$skip_flag`

AzureOrange404 replied:
changed to:
```sh
mpv*) nohup "$player_function" "$skip_flag" "$skip_arg" --force-media-title="${allanime_title}Episode ${ep_no}" "$episode" >/dev/null 2>&1 & ;;
```

synacktraa commented on ani-cli:
```diff
@@ -104,6 +106,17 @@ dep_ch() {
+        skip_script="--script=~/.config/mpv/scripts/skip.lua"
+    fi
+}
+
```
This function calling is not needed anymore.

synacktraa commented on ani-cli:
```diff
@@ -104,6 +106,17 @@ dep_ch() {
 
 # SCRAPING
 
+get_skip_times() {
+    skip_argument="$(ani-skip "$title" "$ep_no")"
+    echo "\"$title\"" "\"$ep_no\"" 1>&2
+    if [ "$skip_argument" = "" ]; then
+        printf "\033[1;31m%s\033[0m Failed\n" "ani-skip" 1>&2
+    else
+        printf "\033[1;32m%s\033[0m Skip Times Fetched\n" "ani-skip" 1>&2
+        skip_script="--script=~/.config/mpv/scripts/skip.lua"
+    fi
+}
```
This function calling is not needed anymore.

Or we can just update it to exit ani-cli, if $skip_flag is null.

AzureOrange404 replied:
Was completely removed.

port19x reviewed (commented):
`ani-cli` should not exit on a failed retrieval of `skip_argument`.
`get_skip_times` may be omitted as per review of @synacktraa

synacktraa commented:
> `ani-cli` should not exit on a failed retrieval of `skip_argument`. `get_skip_times` may be omitted as per review of @synacktraa

Please don't release a new version just after merging @AzureOrange404's code, I want to update the readme before that.

71zenith commented:
@synacktraa pls make the script in the repo an executable to reduce the install steps

synacktraa commented:
> @synacktraa pls make the script in the repo an executable to reduce the install steps

Done. Interesting, I had no idea it could be done like that

AzureOrange404 commented:
> `ani-cli` should not exit on a failed retrieval of `skip_argument`. `get_skip_times` may be omitted as per review of @synacktraa

it is only exiting if the dependency is missing, otherwise it prints a warning, but still runs.
I'll look in to the rest of the suggestions though.

AzureOrange404 commented:
> I have updated it, do tell me If there are any more issues with `ani-skip`.
> 
> Update the ani-cli's `get_skip_times`, as I've updated ani-skip script. So now you'll have to just call ani-skip, it will handle the errors itself.

I don't know if it's just me, but now the player jumps to 0:00 after skipping.
I get this mpv error:
```
[skip] Error: Can't convert '301.499 --script=/home/azure/.config/mpv/scripts/skip.lua' to number!
[skip] script-opts: error converting value '301.499 --script=/home/azure/.config/mpv/scripts/skip.lua' for key 'end_time'
```

synacktraa commented:
> > I have updated it, do tell me If there are any more issues with `ani-skip`.
> 
> > 
> 
> > Update the ani-cli's `get_skip_times`, as I've updated ani-skip script. So now you'll have to just call ani-skip, it will handle the errors itself.
> 
> 
> 
> I don't know if it's just me, but now the player jumps to 0:00 after skipping.
> 
> I get this mpv error:
> 
> ```
> 
> [skip] Error: Can't convert '301.499 --script=/home/azure/.config/mpv/scripts/skip.lua' to number!
> 
> [skip] script-opts: error converting value '301.499 --script=/home/azure/.config/mpv/scripts/skip.lua' for key 'end_time'
> 
> ``` 

Wait, I'll look into it

AzureOrange404 commented:
> > > I have updated it, do tell me If there are any more issues with `ani-skip`.
> > 
> > 
> > > 
> > 
> > 
> > > Update the ani-cli's `get_skip_times`, as I've updated ani-skip script. So now you'll have to just call ani-skip, it will handle the errors itself.
> > 
> > 
> > I don't know if it's just me, but now the player jumps to 0:00 after skipping.
> > I get this mpv error:
> > ```
> > 
> > [skip] Error: Can't convert '301.499 --script=/home/azure/.config/mpv/scripts/skip.lua' to number!
> > 
> > [skip] script-opts: error converting value '301.499 --script=/home/azure/.config/mpv/scripts/skip.lua' for key 'end_time'
> > ```
> 
> Wait, I'll look into it

If mpv is run manually, it works.
It's just the implementation via ani-cli. I guess somehow due to variable handling.

Try:
```sh
mpv "$(ani-skip "Ansatsu Kyoushitsu (TV) (22 episodes)" 1)" --force-media-title="Ansatsu Kyoushitsu Episode 1" https://video.wixstatic.com/video/f5c58e_f62ec058c47d41798e86987023d35063/1080p/mp4/file.mp4
```

synacktraa commented:
> > > > I have updated it, do tell me If there are any more issues with `ani-skip`.
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > Update the ani-cli's `get_skip_times`, as I've updated ani-skip script. So now you'll have to just call ani-skip, it will handle the errors itself.
> > > 
> > > 
> > > I don't know if it's just me, but now the player jumps to 0:00 after skipping.
> > > I get this mpv error:
> > > ```
> > > 
> > > [skip] Error: Can't convert '301.499 --script=/home/azure/.config/mpv/scripts/skip.lua' to number!
> > > 
> > > [skip] script-opts: error converting value '301.499 --script=/home/azure/.config/mpv/scripts/skip.lua' for key 'end_time'
> > > ```
> > 
> > 
> > Wait, I'll look into it
> 
> If mpv is run manually, it works. It's just the implementation via ani-cli. I guess somehow due to variable handling.
> 
> Try:
> 
> ```shell
> mpv "$(ani-skip "Ansatsu Kyoushitsu (TV) (22 episodes)" 1)" https://video.wixstatic.com/video/f5c58e_f62ec058c47d41798e86987023d35063/1080p/mp4/file.mp4
> ```

Can you once show how is it being in your ani-cli?

AzureOrange404 commented:
> > > > > I have updated it, do tell me If there are any more issues with `ani-skip`.
> > > > 
> > > > 
> > > > > 
> > > > 
> > > > 
> > > > > Update the ani-cli's `get_skip_times`, as I've updated ani-skip script. So now you'll have to just call ani-skip, it will handle the errors itself.
> > > > 
> > > > 
> > > > I don't know if it's just me, but now the player jumps to 0:00 after skipping.
> > > > I get this mpv error:
> > > > ```
> > > > 
> > > > [skip] Error: Can't convert '301.499 --script=/home/azure/.config/mpv/scripts/skip.lua' to number!
> > > > 
> > > > [skip] script-opts: error converting value '301.499 --script=/home/azure/.config/mpv/scripts/skip.lua' for key 'end_time'
> > > > ```
> > > 
> > > 
> > > Wait, I'll look into it
> > 
> > 
> > If mpv is run manually, it works. It's just the implementation via ani-cli. I guess somehow due to variable handling.
> > Try:
> > ```shell
> > mpv "$(ani-skip "Ansatsu Kyoushitsu (TV) (22 episodes)" 1)" https://video.wixstatic.com/video/f5c58e_f62ec058c47d41798e86987023d35063/1080p/mp4/file.mp4
> > ```
> 
> Can you once show how is it being in your ani-cli?

I implemeted your changes:
```sh
        mpv*) "$player_function" "$skip_flag" --force-media-title="${allanime_title}Episode ${ep_no}" "$episode" ;;
```

synacktraa commented:
> > Can you once show how is it being in your ani-cli?
> 
> I implemeted your changes:
> 
> ```shell
> mpv*) nohup "$player_function" "$skip_flag" --force-media-title="${allanime_title}Episode ${ep_no}" "$episode" >/dev/null 2>&1 & ;;
> ```

remove quotes from `$skip_flag`, as it has two flags

AzureOrange404 commented:
> 
> remove quotes from `$skip_flag`, as it has two flags

That's the problem, shellcheck will complain if the quotes are removed. 
There is no way (I know of) in POSIX to pass two flags to a command using one variable.

synacktraa commented:
> > remove quotes from `$skip_flag`, as it has two flags
> 
> That's the problem, shellcheck will complain if the quotes are removed. There is no way (I know of) in POSIX to pass two flags to a command using one variable. (That's why I had two variables before, I guess.)

Yeah, I get it. My previous script was not even including the lua script but I thought `ani-skip` script should be the one to return the script and script-opts flag together. let's see what maintainers say and then I'll update ani-skip accordingly.

AzureOrange404 commented:
> 
> Yeah, I get it. My previous script was not even including the lua script but I thought `ani-skip` script should be the one to return the script and script-opts flag together. let's see what maintainers say and then I'll update ani-skip accordingly.

I could just split your output to two variables, would be no problem.

@all 
I just saw that, if there is no output (e. g. anime missing) and the flag is empty '', mpv throws the following error:
```
[file] Cannot open file '': No such file or directory
Failed to open .
```
But if everyone's fine with this, it can just be ignored.

synacktraa commented:
> 
> I could just split your output to two variables, would be no problem.
> 
> **at all** I just saw that, if there is no output (e. g. anime missing) and the flag is empty '', mpv throws the following error:
> 
> ```
> [file] Cannot open file '': No such file or directory
> Failed to open .
> ```
> 
> But if everyone's fine with this, it can just be ignored.



> > Yeah, I get it. My previous script was not even including the lua script but I thought `ani-skip` script should be the one to return the script and script-opts flag together. let's see what maintainers say and then I'll update ani-skip accordingly.
> 
> I could just split your output to two variables, would be no problem.
> 
> **at all** I just saw that, if there is no output (e. g. anime missing) and the flag is empty '', mpv throws the following error:
> 
> ```
> [file] Cannot open file '': No such file or directory
> Failed to open .
> ```
> 
> But if everyone's fine with this, it can just be ignored.

Or we can just call it like this:

```sh
mpv*) nohup "$player_function" --force-media-title="${allanime_title}Episode ${ep_no}" "$episode" "$skip_flag" >/dev/null 2>&1 & ;;
```

AzureOrange404 commented:
> 
> Or we can just call it like this:
> 
> ```shell
> mpv*) nohup "$player_function" --force-media-title="${allanime_title}Episode ${ep_no}" "$episode" "$skip_flag" >/dev/null 2>&1 & ;;
> ```

Yeah, then the errors just appear after the video finished, but it doesn't matter as they are suppressed by `nohup` anyways.

AzureOrange404 commented:
> 
> You already pushed the changes :), I was planning to do something about it.

Oh haha, yeah you still can, I guess.
I'll just annoy everyone with another commit. :)

synacktraa commented:
> > > remove quotes from `$skip_flag`, as it has two flags
> 
> > 
> 
> > That's the problem, shellcheck will complain if the quotes are removed. There is no way (I know of) in POSIX to pass two flags to a command using one variable. (That's why I had two variables before, I guess.)
> 
> 
> 
> That's because this is one of the few instances where you actually want shell word splitting.
> 
> You can add a shellcheck ignore directive as found in a few other places in the script

If quotes can be omitted, it will also get rid of the empty $skip_flag issue

synacktraa commented:
@port19x Is this the correct shellcheck?
`# shellcheck disable=SC2086`

CoolnsX commented:
> @port19x Is this the correct shellcheck? `# shellcheck disable=SC2086`

yes

synacktraa commented:
> if your shellcheck complains, you can run shellcheck outside of script like this --
> 
> ```
> shellcheck ani-cli
> ```
> 
> it will spit out warnings/errors and also those SCXXXX error codes, just copy and paste them just above the line of code where you want to disable it.

@AzureOrange404 Update it accordingly, you should have waited for the solution before committing :>

synacktraa commented on ani-cli:
```diff
@@ -237,13 +239,14 @@ download() {
             printf "%s\n" "$episode"
             ;;
-        mpv*) nohup "$player_function" --force-media-title="${allanime_title}Episode ${ep_no}" "$episode" >/dev/null 2>&1 & ;;
+        mpv*) nohup "$player_function" "$skip_flag" "$skip_arg" --force-media-title="${allanime_title}Episode ${ep_no}" "$episode" >/dev/null 2>&1 & ;;
```
It should be:
```sh
# shellcheck disable=SC2086
mpv*) nohup "$player_function" $skip_flag --force-media-title="${allanime_title}Episode ${ep_no}" "$episode" >/dev/null 2>&1 & ;;
```

I hope this is right @CoolnsX

synacktraa replied:
@AzureOrange404 please verify it once.

AzureOrange404 replied:
done and working

AzureOrange404 commented on ani-cli:
```diff
@@ -237,13 +239,14 @@ download() {
 }
 
 play_episode() {
+    [ "$skip_intro" = 1 ] && skip_flag="$(ani-skip "$title" "$ep_no" | cut -d' ' -f1)" && skip_arg="$(ani-skip "$title" "$ep_no" | cut -d' ' -f2)"
```
done

AzureOrange404 commented:
> 
> That's because this is one of the few instances where you actually want shell word splitting. You can add a shellcheck ignore directive as found in a few other places in the script

Didn't know, didn't look it up.
But now it's implemented and it works, thanks.

> 
> @AzureOrange404 Update it accordingly, you should have waited for the solution before committing :>

Yeah, maybe I should have, haha.

AzureOrange404 commented:
@synacktraa Could you change `printf` in ani-skip line 37 to `die`? Otherwise it'll still output. ;)

justchokingaround commented:
also why tf are we using an external script with `ani-skip`. i get why the lua script is needed, but why call another script inside of ani-cli ?

(this can probably be answered if i backread the whole thread, but i won't be doing that)

synacktraa commented:
> @synacktraa Could you change `printf` in ani-skip line 37 to `die`? Otherwise it'll still output. ;)

Yeah, what a stupid mistake

port19x commented:
> also why tf are we using an external script with `ani-skip`. i get why the lua script is needed, but why call another script inside of ani-cli ?
> 
> (this can probably be answered if i backread the whole thread, but i won't be doing that)

Because it's a preexisting solution from what I've gathered and we delegate responsibility and locs that way

port19x commented:
Perhaps tomorrow I'll finally get around to the mac testing I promised

71zenith commented:
Since we are using the unquoted variable approach, it would be much better to implement a proper player argument option and integrate `ani-skip` in it. however that can be left to `4.7.x`

71zenith commented:
@synacktraa can u state a testcase anime and episode for ani-skip

71zenith commented:
@synacktraa also pls implement skipping for outro as well

71zenith commented:
modify the script to redirect stderr if the script cannot find the anime
![image](https://github.com/pystardust/ani-cli/assets/92977828/010f1ec2-7dcd-4124-8c78-cac4fa37db34)
dont rely on a edge case

71zenith commented:
![image](https://github.com/pystardust/ani-cli/assets/92977828/25179b8c-4dc3-490c-8219-141e7b35b9e2)
did none of u check if this runs, modify ani-cli or ani-skip to discard the `(n episodes)`
MAL is already extremely picky and does not have the best search engine, it will never give the right results if we add unnecessary metadata to the query

synacktraa commented:
> modify the script to redirect stderr if the script cannot find the anime
> 
> ![image](https://github.com/pystardust/ani-cli/assets/92977828/010f1ec2-7dcd-4124-8c78-cac4fa37db34)
> 
> dont rely on a edge case

Isn't die function already redirecting it to stderr?

71zenith commented:
my bad, it looks like it does discard the (n episodes) however it seems that u try to match the exact title name in the mal results. this i doubt will work for anime with >1 seasons and those which are named differently in allanime

synacktraa commented:
> > ![image](https://github.com/pystardust/ani-cli/assets/92977828/25179b8c-4dc3-490c-8219-141e7b35b9e2)
> > 
> > did none of u check if this runs, modify ani-cli or ani-skip to discard the `(n episodes)`
> > 
> > MAL is already extremely picky and does not have the best search engine, it will never give the right results if we add unnecessary metadata to the query
> 
> I haven't tested ani-skip with ani-cli yet, been just resolving ani-skip issues

Tell me if you're modifying ani-cli, else I'll do something with ani-skip

synacktraa commented:
Someone will have to first take me through `ani-cli` code, especially the episode range parsing thing.

port19x commented on ani-cli:
```diff
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-version_number="4.6.2"
+version_number="4.7"
```
```suggestion
version_number="4.7.0"
```

AzureOrange404 replied:
sorry, my bad, is done.

port19x reviewed (changes requested):
3 digit version number please

synacktraa commented:
> @synacktraa also pls implement skipping for outro as well

Here you go. https://github.com/synacktraa/ani-skip/commit/657a1de142ed1f7f846b130540cbb4564f7c42fa

port19x commented:
I plan to merge and release sometime later today

port19x commented:
Looks good and works on my end

Needs sign off from @Derisis13 due to requested changes

Derisis13 commented on README.md:
```diff
@@ -401,6 +402,31 @@ apk del grep sed curl fzf git aria2 ffmpeg ncurses
+
+By setting the environment variable `ANI_CLI_SKIP_INTRO=1`, the feature can be enabled permanently.
+
+#### Install
```
The entire install section could be omitted in favor of a link to their readme's install section

AzureOrange404 replied:
I removed the install section, and pointed to the original ani-skip readme.
Never mind the going back and forth.

Derisis13 commented on README.md:
```diff
@@ -401,6 +402,31 @@ apk del grep sed curl fzf git aria2 ffmpeg ncurses
 - ffmpeg - m3u8 Downloader (fallback)
 - fzf - User interface
 
+### Ani-Skip (optional)
```
it should have a bullet point for good measure similarly to ffmpeg and the others

AzureOrange404 replied:
I added it to the list.
Can I leave the ani-skip section as it is now?

Derisis13 commented on README.md:
```diff
@@ -401,6 +402,31 @@ apk del grep sed curl fzf git aria2 ffmpeg ncurses
+
+By using the `--skip` flag, ani-cli runs the external ani-skip script if it is installed (if the dependency check fails, ani-cli exits).
+
+By setting the environment variable `ANI_CLI_SKIP_INTRO=1`, the feature can be enabled permanently.
```
we haven't documented usage in readme for a long time, and since you have it in the manpage, this can be taken out too

AzureOrange404 replied:
I removed the usage section as well.
Someone requested it to be in the readme if I remember right. ;)

lasersPew commented:
Imma test some stuff with windows and WSL.

synacktraa commented:
> Imma test some stuff with windows and WSL.


ani-skip is working in Wsl
Try gitbash too if possible

synacktraa commented:
> I am on Manjaro, not sure if that affects but getting the following output for ani-skip and one piece, am I doing something wrong? I tried the same with this PR as well but didn't see any skip so tried to do this.
> 
> ```
> 
> ani-skip "1P (1084 episodes)" 341
> 
> --script-opts= '--script=/home/user/.config/mpv/scripts/skip.lua'
> 
> ```

I'll look into it.

lasersPew commented:
> > I am on Manjaro, not sure if that affects but getting the following output for ani-skip and one piece, am I doing something wrong? I tried the same with this PR as well but didn't see any skip so tried to do this.
> > ```
> > 
> > ani-skip "1P (1084 episodes)" 341
> > 
> > --script-opts= '--script=/home/user/.config/mpv/scripts/skip.lua'
> > ```
> 
> ![image](https://user-images.githubusercontent.com/91981716/281834577-607481c4-4d75-4ce1-8bd8-1e3ae40b8d23.png)
> 
> update ani-skip once. `sudo ani-skip -U`

This actually doesn't solve the problem. tried using `One Piece` instead of `1P (1084 episodes)` and got the script opts:
![image](https://github.com/pystardust/ani-cli/assets/87552756/35c16cda-8163-40a4-8a29-8063bd62e114)

71zenith commented:
works nicely (would have preferred to run it in background)
about the one piece thing. im afraid there is so much we can do about it. they likely do it to not get noticed by the big corp daddies.

synacktraa commented:
> This actually doesn't solve the problem. tried using `One Piece` instead of `1P (1084 episodes)` and got the script opts: ![image](https://user-images.githubusercontent.com/87552756/283069790-35c16cda-8163-40a4-8a29-8063bd62e114.png)

If there was no error, it means there was an entry related to `1P`, it's just that parsing is not working well. I'll do something about it. If it was python, we could have used vectors or ngrams to get most relevant entry.

NKTMQ commented:
> > Imma test some stuff with windows and WSL.
> 
> ani-skip is working in Wsl Try gitbash too if possible

For any windows users that setup ani-cli with scoop, the likely best way is to use scoop to create a shim for ani-skip as well.

step 1
`scoop install { "version": "1.0", "url": "https://raw.githubusercontent.com/synacktraa/ani-skip/master/ani-skip", "bin": "ani-skip" }`

step 2
Place [https://raw.githubusercontent.com/synacktraa/ani-skip/master/skip.lua](https://raw.githubusercontent.com/synacktraa/ani-skip/master/skip.lua)
Into the mpv scripts folder. 

Scripts folder autoruns any lua files in it so we can ignore the "--script=$HOME/.config/mpv/scripts/skip.lua" component of the ani-skip output as mpv will ignore it (with an error message saying file not found).

Otherwise if you'd like to have the script elsewhere you can edit the ani-skip's 4th line from the scoop install to be a windows folder path instead.
